src/LAoP/Matrix/Alternative.hs

```diff
@@ -0,0 +1,157 @@
+instance Num e => Category (Matrix e) where
+    id = Identity
+
+    Identity   . x          = x
+    x          . Identity   = x
+    Zero       . _          = Zero
+    _          . Zero       = Zero
+    Lift f x y . z          = Lift f (x . z) (y . z)
+    x          . Lift f y z = Lift f (x . y) (x . z)
+    Join w x   . Fork y z   = (w . y) + (x . z)
+    Fork x y   . z          = Fork (x . z) (y . z)
+    x          . Join y z   = Join (x . y) (x . z)
+
+-- Adapted from https://hackage.haskell.org/package/categories
+class Category k => Cartesian k where
+    type Product k :: * -> * -> *
+    fst   :: Product k a b `k` a
+    snd   :: Product k a b `k` b
+    (&&&) :: (a `k` b) -> (a `k` c) -> a `k` Product k b c
+
+instance Cartesian (->) where
+    type Product (->) = (,)
+    fst = Prelude.fst
+    snd = Prelude.snd
+    (f &&& g) a = (f a, g a)
+
+instance Num e => Cartesian (Matrix e) where
+    type Product (Matrix e) = Either
+    fst   = Join Identity Zero
+    snd   = Join Zero Identity
+    (&&&) = Fork
+
+-- A standard construction for any Cartesian category.
+bimapProduct :: Cartesian k => k a c -> k b d -> Product k a b `k` Product k c d
+bimapProduct f g = (f . fst) &&& (g . snd)
+
+-- For free!
+(><) :: Num e => Matrix e a b -> Matrix e c d -> Matrix e (Either a c) (Either b d)
+(><) = bimapProduct
+
+infixl 4 ><
+
+class Category k => CoCartesian k where
+    type Sum k :: * -> * -> *
+    inl   :: a `k` Sum k a b
+    inr   :: b `k` Sum k a b
+    (|||) :: k a c -> k b c -> Sum k a b `k` c
+
+-- A standard construction for any CoCartesian category.
+bimapSum :: CoCartesian k => k a c -> k b d -> Sum k a b `k` Sum k c d
+bimapSum f g = (inl . f) ||| (inr . g)
+
+-- For free!
+(-|-) :: Num e => Matrix e a b -> Matrix e c d -> Matrix e (Either a c) (Either b d)
+(-|-) = bimapSum
+
+infixl 5 -|-
+
+instance CoCartesian (->) where
+    type Sum (->) = Either
+    inl = Left
+    inr = Right
+    (f ||| _) (Left  a) = f a
+    (_ ||| g) (Right a) = g a
+
+instance Num e => CoCartesian (Matrix e) where
+    type Sum (Matrix e) = Either
+    inl = Fork Identity Zero
+    inr = Fork Zero Identity
+    (|||) = Join
+
+class (Cartesian k, CoCartesian k) => Distributive k where
+    distribute :: Product k a (Sum k b c) `k` Sum k (Product k a b) (Product k a c)
+
+instance Distributive (->) where
+    distribute (a, Left  b) = Left  (a, b)
+    distribute (a, Right c) = Right (a, c)
+
+instance Num e => Distributive (Matrix e) where
+    distribute = Fork (id -|- fst) (id -|- snd)
+
+transpose :: Matrix e a b -> Matrix e b a
+transpose m = case m of
+    Identity   -> Identity
+    Zero       -> Zero
+    Lift f x y -> Lift f (transpose x) (transpose y)
+    Join x y   -> Fork (transpose x) (transpose y)
+    Fork x y   -> Join (transpose x) (transpose y)
+
+select :: Num e => Matrix e a (Either b c) -> Matrix e b c -> Matrix e a c
+select x y = Join y id . x
+
+kr :: Num e => Matrix e a b -> Matrix e a c -> Matrix e a (Either b c)
```

Comment:
Khatri Rao type should be `kr :: Num e => Matrix e a b -> Matrix e a c -> Matrix e a (b, c)`. Matrix has a Monoidal structure! That's why I said it might be tricky to implement without a constructor.

References: 
- https://pdfs.semanticscholar.org/ccf5/27fa9179081223bffe8067edd81948644fc0.pdf (page 12)
- https://haslab.uminho.pt/jno/files/scp11.pdf (page 38)

Reply:
Oops, you're right! I got carried away replacing `fstM` and `sndM` to them `M`-less versions :)

Reply:
No problem! :smile:  Probably another constructor will solve the issue, but I'm not sure about the projections.